test/diag2dwd.jl

```diff
@@ -217,6 +218,317 @@ end
     @test nparts(advdiffdp, :Op2) == 2
 end
 
+@testset "Decapode Composition" begin
+
+  """
+      function unique_by!(acset, column_names::Vector{Symbol})
+
+  Given column names from the same table, remove duplicate rows.
+  
+  WARNING: This function does not check if other tables index into the one
+  given. Removal of rows is performed with prejudice.
+
+  See also: [`unique_by`](@ref).
+
+  # Examples
+  ```julia-repl
+  julia> g = @acset Graph begin
+    V = 2
+    E = 3
+    src = [1,1,2]
+    tgt = [1,1,2]
+  end
+  julia> unique_by!(g, :E, [:src, :tgt])
+  ``` 
+  """
+  function unique_by!(acset, table::Symbol, columns::Vector{Symbol})
+    rows = columns .|> (x -> acset[x]) |> (x -> zip(x...)) |> collect
+    rem_parts!(acset, table,
+      setdiff(parts(acset, table),
+        unique(i -> rows[i], eachindex(rows))))
+    return acset
+  end 
+
+  """
+      function unique_by(acset, column_names::Vector{Symbol})
+
+  Given column names from the same table, return a copy of the acset with
+  duplicate rows removed. Removal of rows is performed with prejudice.
+
+  WARNING: This function does not check if other tables index into the one
+  given. Removal of rows is performed with prejudice.
+
+  See also: [`unique_by!`](@ref).
+
+  # Examples
+  ```julia-repl
+  julia> g = @acset Graph begin
+    V = 2
+    E = 3
+    src = [1,1,2]
+    tgt = [1,1,2]
+  end
+  julia> unique_by(g, :E, [:src, :tgt])
+  ``` 
+  """
+  function unique_by(acset, table::Symbol, columns::Vector{Symbol})
+    acset_copy = copy(acset)
+    unique_by!(acset_copy, table, columns)
+  end 
+
+  """
+      function compose_decapodes(decapodes_vars::Vector{Tuple{NamedDecapode{
+        Any, Any, Symbol}, Vector{Symbol}}}, relation::RelationDiagram)
+
+  Compose a list of decapodes as specified by the given relation diagram.
+
+  The decapodes must be given in the same order as they were specified in the
+  relation.
+
+  State variables (such as the (C,V) given in the head of the following
+  @relation) do not affect the result of a composition.
+
+  # Examples
+  ```julia-repl
+  julia> compose_diff_adv = @relation (C,V) begin
+    diffusion(C, ϕ₁)
+    advection(C, ϕ₂, V)
+    superposition(ϕ₁, ϕ₂, ϕ, C)
+  end
+  julia> compose_decapodes([(Diffusion, [:C, :ϕ]), (Advection, [:C, :ϕ, :V]),
+    (Superposition, [:ϕ₁, :ϕ₂, :ϕ, :C])], compose_diff_adv)
+  ````
+  """
+  function compose_decapodes(decapodes_vars::Vector{Tuple{NamedDecapode{
+    Any, Any, Symbol}, Vector{Symbol}}}, relation::RelationDiagram)
+    # TODO: Calling this function "compose" is something of a misnomer.  e.g.
+    # You could very well give a vector containing a single decapode. Then the
+    # effect of applying the relation would be something like renaming
+    # variables. So this function should really be called oapply_decapodes (or
+    # just oapply if you are willing to import oapply from
+    # Catlab.WiringDiagrams).
+    r = relation
+    copies = decapodes_vars .|> first .|> copy
```

Comment:
the broadcasted `|>` is not really idiomatic. This syntax is more for shell-style scripts rather than real library code.